resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -60,6 +62,43 @@ private[spark] class KubernetesClusterSchedulerBackend(
     removeExecutor(executorId, reason)
   }
 
+  /**
+    * Get an application ID associated with the job.
+    * This returns the string value of [[appId]] if set, otherwise
+    * the locally-generated ID from the superclass.
+    * @return The application ID
+    */
+
+  var appId: Option[String] = None;
+
+  override def applicationId(): String = {
+
+    appId.map(_.toString).getOrElse {
+      logInfo("Initializing Application ID.")
+      bindApplicationId();
+      appId.get
+    }
+  }
+
+  def bindApplicationId(): Unit = {
+    val appIdString = {
+      val wasSparkSubmittedInClusterMode = conf.get(KUBERNETES_DRIVER_SUBMIT_CHECK)
+
+      // cluster mode: get appId from driver env
+      if (wasSparkSubmittedInClusterMode) {
```

Comment:
Instead of checking `wasSparkSubmittedInClusterMode`, I think you can simply check `spark.app.id` and use the value of it if it's set and use `super.applicationId`. In client mode, since `KubernetesClientApplication` is not invoked, how the application ID is generated doesn't really matter, so using `super.applicationId` is fine for client mode. So the logic is simple: if `spark.app.id` is set, use the value of it, otherwise use `super.applicationId`.